### ISSUE TITLE
Support for repeating ImageTextures

### ISSUE THREAD

**Issue by swifmaneum** [bartmcleod/VrmlParser#11]
Hi there :)

I was wondering if VrmlParser already supports repeating ImageTextures? Explicitly I'm thinking of the `repeatS` and `repeatT` properties, which are described [here](http://www.c3.hu/cryptogram/vrmltut/part6.html): 

> The repeatS and repeatT fields specify how the texture wraps in the S and T directions. ... If repeatS is TRUE (the default), the texture map is repeated outside the [0.0, 1.0] texture coordinate range in the S direction so that it fills the shape. If repeatS is FALSE, the texture coordinates are clamped in the S direction to lie within the [0.0, 1.0] range. The repeatT field is analogous to the repeatS field.

I was testing with a ImageTexture like so:
```
texture ImageTexture
{
    url "assets/wrl/textures/wood.png"
    repeatS	TRUE
    repeatT	TRUE
}
```

But changing `repeatS` and `repeatT` to `FALSE` did not seem to look any different.

**Comment by bartmcleod** [comment]
From the top of my mind I would says these properties have no effect yet. I am keen on supporting them though.

**Comment by bartmcleod** [comment]
@swifmaneum I started work on it last night. The specification is not overly clear on how textures should wrap, especially not in the case of standard objects, but I made some progress. It is also a bit hard to translate the wrapping rules to how wrapping is done in ThreeJs. It is even harder to figure out the rules for repetition of textures and how they should scale when they repeat, but I'm on it.

To summarize: It is easy to make textures repeat, but it is harder to comply with the VRML 97 rules for wrapping and repetition.

**Comment by swifmaneum** [comment]
Cool :) 
Hmm I did not fully understand the specification either. For me it would not be important to comply the spec, I would be perfectly happy if my texture would be repeated, once I set `repeatS` and `repeatT` to `TRUE`.

**Comment by bartmcleod** [comment]
Yes, I understand that. The spec are deprecated anyway, but I should try to stick to them as close as possible, within reason. What I find hard to distill from the spec is the number of repetions. From what I understood so far, a textCoord node should be provided, so that the wrapping of the texture can be done in a way defined by the textCoord node. Once I can find how this would apply to a shape such as a box, then I can seek to translate it to three.js. For now, the number of repitions would be a mere guess. Would it be possible for you to craft a simple example wrl with a proposed repetition and maybe a graphical impression of what you think that should look like when rendered?

**Comment by bartmcleod** [comment]
![image](https://user-images.githubusercontent.com/1441397/55034697-5b227c00-5016-11e9-90a3-f95003966267.png)

**Comment by swifmaneum** [comment]
I wasn't aware that you have to specify the number of repetitions manually in Three.js as in the example [here](https://threejs.org/docs/#api/en/textures/Texture.repeat). I thought the texture would be repeated until the 'textured object ends'. Maybe it is possible to calculate the number of repetitions needed somewhat like: 

`numberOfRepetitions = texturedObject.width / texture.image.width`

I can craft a little example if you still need one. But he screenshot you uploaded looks a lot like what I want :)

**Comment by bartmcleod** [comment]
The algorithm you suggest would yield a repetion of 1 if a standard size texture image of 256 x 256 px is used ;-)

The VRML spec allows you to add a textCoord property with a TextureCoordinate node with ranges of vertices on your object. If repeat is set, then the texture repetition would extend beyond the 0,1 range in the textCoord. This only makes sense on irregularly shaped objects, like IndexedFaceSet. For a box that you would want to have 4 tiles in the t direction, it would mean you would devide the x side of the box into five coordinates, so that 4 tiles can be counted. That said, the count would simply be the number of coordinates -1... Something along those lines should work. Of course it is also possible to do things that are illegal in VRML, such as adding new properties to a Texture node and using those. For now, this is just a bit of guessing. People who can help clarify this are encouraged to comment.

**Comment by swifmaneum** [comment]
Following example might show the issue pretty well. I'm using a chessboard like texture to cover a IndexedFaceSet, but instead of repeating the chessboard, the texture will be stretched:
![textureNotRepeating](https://user-images.githubusercontent.com/7429598/55068868-11bc4600-5083-11e9-88fb-63794d0b60c5.PNG)

I'd expect the plane to be covered with the chessboard.

I'm using following Image for my ImageTexture: 
![checkerboard](https://user-images.githubusercontent.com/7429598/55069203-ca828500-5083-11e9-893d-7ba7741a866a.jpg)

And following VRML (again it is pretty long as it is mostly generated): 

````
#VRML V2.0 utf8

#Cosmo Worlds V2.0

DEF The_Start Viewpoint {
position 1397.84 1700 -2862.15
orientation -0.408248 0.53204 0.169102 0.987861
fieldOfView 0.8
description "Start"
}
DEF wange Transform{
translation 0 20 0
rotation 2 0 0 -1.5708
children[
DEF imos Transform{
children[
DEF R49153 Transform{
children[
DEF R49154 Transform{
children[
DEF R16385 Transform{
translation 25 0 0
rotation 0 2 0 -1.5707963
children[
Shape{
geometry IndexedFaceSet{
coord DEF BT16385CO Coordinate{
point[
0 2 0,
0 2 19,
0 2 0,
0 0 0,
0 0 19,
0 2 19,
2200 2 0,
2200 0 0,
2200 0 19,
2200 2 19,
2200 2 0,
2200 2 19,
2200 605 0,
2200 605 19,
0 605 0,
0 605 19,
0 2 0,
0 2 19,
]
}
}
}
DEF Surf0 Transform{children[Shape{
appearance DEF APP_UNDEFINED Appearance{
material Material{
ambientIntensity 0.4
diffuseColor 0.8 0.8 0.8
specularColor 0.4 0.4 0.4
transparency 0
shininess 0.2
#reflectivity 0.5
}
texture ImageTexture
{
url "assets/wrl/textures/checkerboard.jpg"
repeatS	TRUE
repeatT	TRUE
}
}
geometry IndexedFaceSet{
coord USE BT16385CO
coordIndex[
10,16,14,-1,
14,12,10,-1,
]
texCoord TextureCoordinate{
point[
7.33333 0.00666667, 0 0.00666667, 0 2.01667
, 0 2.01667, 7.33333 2.01667, 7.33333 0.00666667
]
}
texCoordIndex[
0, 1, 2, -1
, 3, 4, 5, -1
]
}
}]}
DEF Surf1 Transform{children[Shape{
appearance DEF APPR Appearance{
material Material{
ambientIntensity 0.4
diffuseColor 0.8 0.8 0.8
specularColor 0.4 0.4 0.4
transparency 0
shininess 0.2
#reflectivity 0.5
}
texture ImageTexture
{
url "assets/wrl/textures/white.png"
repeatS	TRUE
repeatT	TRUE
}
}
geometry IndexedFaceSet{
coord USE BT16385CO
coordIndex[
13,15,17,-1,
17,11,13,-1,
]
texCoord TextureCoordinate{
point[
7.33333 2.01667, 0 2.01667, 0 0.00666667
, 0 0.00666667, 7.33333 0.00666667, 7.33333 2.01667
]
}
texCoordIndex[
0, 1, 2, -1
, 3, 4, 5, -1
]
}
}]}
DEF Edge1 Transform{children[Shape{
appearance USE APPR
geometry IndexedFaceSet{
convex FALSE
coord USE BT16385CO
coordIndex[
2,3,4,1,0,-1,
6,7,3,2,-1,
7,8,4,3,-1,
8,9,5,4,-1,
10,11,8,7,6,-1,
]
texCoord TextureCoordinate{
point[
0 0, 0 0, 0 0.0633333, 0 0.0633333, 0 0
, 7.33334 0, 7.33334 0, 0 0, 0 0
, 7.33334 0, 7.33334 0.0633333, 0 0.0633333, 0 0
, 7.33334 0.0633333, 7.33334 0.0633333, 0 0.0633333, 0 0.0633333
, 14.6667 0, 14.6667 0.0633333, 7.33334 0.0633333, 7.33334 0, 7.33334 0
]
}
texCoordIndex[
0, 1, 2, 3, 4, -1
, 5, 6, 7, 8, -1
, 9, 10, 11, 12, -1
, 13, 14, 15, 16, -1
, 17, 18, 19, 20, 21, -1
]
}
}]}
DEF Edge2 Transform{children[Shape{
appearance USE APPR
geometry IndexedFaceSet{
convex FALSE
coord USE BT16385CO
coordIndex[
12,13,11,10,-1,
]
texCoord TextureCoordinate{
point[
0 0, 0 0.0633333, 0 0.0633333, 0 0
]
}
texCoordIndex[
0, 1, 2, 3, -1
]
}
}]}
DEF Edge3 Transform{children[Shape{
appearance USE APPR
geometry IndexedFaceSet{
convex FALSE
coord USE BT16385CO
coordIndex[
14,15,13,12,-1,
]
texCoord TextureCoordinate{
point[
0 0, 0 0.0633333, 0 0.0633333, 0 0
]
}
texCoordIndex[
0, 1, 2, 3, -1
]
}
}]}
DEF Edge4 Transform{children[Shape{
appearance USE APPR
geometry IndexedFaceSet{
convex FALSE
coord USE BT16385CO
coordIndex[
16,17,15,14,-1,
]
texCoord TextureCoordinate{
point[
0 0, 0 0.0633333, 0 0.0633333, 0 0
]
}
texCoordIndex[
0, 1, 2, 3, -1
]
}
}]}
]
}
]
}
]
}
]
}
]
}

**Comment by bartmcleod** [comment]
Thank you very much for the example! This will likely help the implementation, because TextureCoordinates are already in place. You do realize the image would only need four squares, right? And it could also be implemented using a ColorTexture..

**Comment by swifmaneum** [comment]
Yeah, the example is a little made up. Actually I'm trying to use a wood like texture but I felt like the chessboard texture demonstrates the issue very well.

**Comment by bartmcleod** [comment]
@swifmaneum I have published the texture support today as v0.8.4, but only in the limited sense as demonstrated in the screenshot above. The testfile that produces the scene in the screenshot is also included in that version (also current master).

**Comment by swifmaneum** [comment]
Cool, thanks a lot! I'll close this one.

**Comment by bartmcleod** [comment]
@swifmaneum I have slightly improved texture support and also support for repeating textures. The repitition has been slightly improved, but it will also have changed a bit, depending on how it is used.
![image](https://user-images.githubusercontent.com/1441397/98175370-bd33aa80-1ef6-11eb-8b4e-54e96c37033b.png)
